### ISSUE TITLE
Training of Imagenet with inception-bn in python gets good training accuracy but low validation after epoch 1

### ISSUE THREAD

**Issue by hoaphumanoid** [dmlc/mxnet#3424]
I'm training Imagenet with inception-bn in python:

```bash
python train_imagenet.py --batch-size 144 --lr 0.05 --lr-factor .94 --gpus 0,1,2,3 --num-epoch 60 --network inception-bn
```
In epoch 0 the training and validation accuracy are reasonable, but after that the validation accuracy drops:
```bash
2016-09-29 20:56:15,427 Node[0] Update[8897]: Change learning rate to 4.70000e-02
2016-09-29 20:56:16,214 Node[0] Epoch[0] Resetting Data Iterator
2016-09-29 20:56:16,217 Node[0] Epoch[0] Time cost=7861.295
2016-09-29 20:56:17,215 Node[0] Saved checkpoint to "inception-bn-python-0-0001.params"
2016-09-29 20:58:03,494 Node[0] Epoch[0] Validation-accuracy=0.228867
2016-09-29 20:58:03,495 Node[0] Epoch[0] Validation-top_k_accuracy_5=0.464500
2016-09-29 20:58:03,495 Node[0] Epoch[0] Validation-top_k_accuracy_10=0.573894
2016-09-29 20:58:03,495 Node[0] Epoch[0] Validation-top_k_accuracy_20=0.682352
2016-09-29 20:58:47,169 Node[0] Epoch[1] Batch [50]     Speed: 166.24 samples/sec       Train-accuracy=0.354028
2016-09-29 20:58:47,169 Node[0] Epoch[1] Batch [50]     Speed: 166.24 samples/sec       Train-top_k_accuracy_5=      0.607500
2016-09-29 20:58:47,170 Node[0] Epoch[1] Batch [50]     Speed: 166.24 samples/sec       Train-top_k_accuracy_10      =0.706944
2016-09-29 20:58:47,170 Node[0] Epoch[1] Batch [50]     Speed: 166.24 samples/sec       Train-top_k_accuracy_20      =0.791667
...
...
2016-09-29 23:08:14,484 Node[0] Epoch[1] Batch [8850]   Speed: 162.40 samples/sec       Train-accuracy=0.455139
2016-09-29 23:08:14,484 Node[0] Epoch[1] Batch [8850]   Speed: 162.40 samples/sec       Train-top_k_accuracy_5=      0.713472
2016-09-29 23:08:14,484 Node[0] Epoch[1] Batch [8850]   Speed: 162.40 samples/sec       Train-top_k_accuracy_10      =0.795833
2016-09-29 23:08:14,484 Node[0] Epoch[1] Batch [8850]   Speed: 162.40 samples/sec       Train-top_k_accuracy_20      =0.862917
2016-09-29 23:08:53,043 Node[0] Update[17793]: Change learning rate to 4.41800e-02
2016-09-29 23:08:54,986 Node[0] Epoch[1] Resetting Data Iterator
2016-09-29 23:08:54,990 Node[0] Epoch[1] Time cost=7851.495
2016-09-29 23:08:56,070 Node[0] Saved checkpoint to "inception-bn-python-0-0002.params"
2016-09-29 23:10:42,088 Node[0] Epoch[1] Validation-accuracy=0.069464
2016-09-29 23:10:42,088 Node[0] Epoch[1] Validation-top_k_accuracy_5=0.181394
2016-09-29 23:10:42,088 Node[0] Epoch[1] Validation-top_k_accuracy_10=0.258162
2016-09-29 23:10:42,088 Node[0] Epoch[1] Validation-top_k_accuracy_20=0.355344
2016-09-29 23:11:26,427 Node[0] Epoch[2] Batch [50]     Speed: 164.00 samples/sec       Train-accuracy=0.466111
2016-09-29 23:11:26,428 Node[0] Epoch[2] Batch [50]     Speed: 164.00 samples/sec       Train-top_k_accuracy_5=      0.723194
2016-09-29 23:11:26,428 Node[0] Epoch[2] Batch [50]     Speed: 164.00 samples/sec       Train-top_k_accuracy_10      =0.806944
2016-09-29 23:11:26,428 Node[0] Epoch[2] Batch [50]     Speed: 164.00 samples/sec       Train-top_k_accuracy_20      =0.869444
...
...
2016-09-30 14:30:36,915 Node[0] Epoch[9] Batch [8850]   Speed: 216.61 samples/sec       Train-accuracy=0.678056
2016-09-30 14:30:36,915 Node[0] Epoch[9] Batch [8850]   Speed: 216.61 samples/sec       Train-top_k_accuracy_5=      0.876250
2016-09-30 14:30:36,915 Node[0] Epoch[9] Batch [8850]   Speed: 216.61 samples/sec       Train-top_k_accuracy_10      =0.920833
2016-09-30 14:30:36,916 Node[0] Epoch[9] Batch [8850]   Speed: 216.61 samples/sec       Train-top_k_accuracy_20      =0.952361
2016-09-30 14:31:01,541 Node[0] Update[88961]: Change learning rate to 2.69308e-02
2016-09-30 14:31:08,212 Node[0] Epoch[9] Resetting Data Iterator
2016-09-30 14:31:08,212 Node[0] Epoch[9] Time cost=5910.135
2016-09-30 14:31:08,754 Node[0] Saved checkpoint to "inception-bn-python-0-0010.params"
2016-09-30 14:32:28,038 Node[0] Epoch[9] Validation-accuracy=0.000998
2016-09-30 14:32:28,038 Node[0] Epoch[9] Validation-top_k_accuracy_5=0.005009
2016-09-30 14:32:28,039 Node[0] Epoch[9] Validation-top_k_accuracy_10=0.010317
2016-09-30 14:32:28,039 Node[0] Epoch[9] Validation-top_k_accuracy_20=0.019836
```

Any suggestion?

**Comment by piiswrong** [comment]
is this newest master?

**Comment by hoaphumanoid** [comment]
From [Sept 22](https://github.com/dmlc/mxnet/commit/2d0e3ac8f017b15abf171f7acf0a3631fc4e2970) There has been any substantial change since then?

**Comment by hoaphumanoid** [comment]
There is a bug with the current version of the library. If you roll back to an older version, previosly to this [PR](https://github.com/dmlc/mxnet/pull/2585), then it is working. I used the commit `c3edd8d75c0b9221c2a808da9b755fad1e27249e` and it is working. There is a difference in inception-bn with the padding:
Old one:
`pooling = mx.symbol.Pooling(data=data, kernel=(3, 3), stride=(2, 2), pool_type="max", name=('max_pool_%s_pool' % name))`
[New](https://github.com/dmlc/mxnet/blob/master/example/image-classification/symbol_inception-bn.R#L64) one:
`pooling = mx.symbol.Pooling(data=data, kernel=(3, 3), stride=(2, 2), pad=(1, 1), pool_type="max", name=('max_pool_%s_pool' % name))`

This might be related to this [issue](https://github.com/dmlc/mxnet/issues/3359).

**Comment by hoaphumanoid** [comment]
I confirmed that the problem is not the padding it is something else. If you remove the padding in the last version of the library you get the following error that has to do with the internal sizes of the matrixes:

````bash
`InferShape Error in ch_concat_3c_chconcat: [09:25:44] src/operator/./concat-inl.h:152: Check failed: (dshape[j]) == (tmp[j]) Incorrect shape[2]: (576,320,13,13). (first input shape: (576,576,14,14))
```
If you roll back to commit `c3edd8d75c0b9221c2a808da9b755fad1e27249e ` the error dispears

**Comment by George19890617** [comment]
I do the same experiment, I once got top1 near 50% on validation set, but now retrain fail to near 13%..., did you know what is the problem?

**Comment by winstywang** [comment]
Stay tuned. We have already trained a new Inception-BN model. Will make it public soon.

**Comment by George19890617** [comment]
@winstywang @piiswrong 
I check the model trained in old and new.
the new:

> 
In [7]: gamma = model.arg_params['bn_5b_3x3_reduce_gamma']

In [8]: gamma
Out[8]: <mxnet.ndarray.NDArray at 0x7fa2c2c59390>

In [9]: gamma.asnumpy()
Out[9]: 
array([ 0.49560896,  0.49560896,  0.49560896,  0.49560896,  ....

the old: 

> In [6]: gamma = model.arg_params['bn_5b_3x3_reduce_gamma']

In [7]: gamma
Out[7]: <mxnet.ndarray.NDArray at 0x7fc48803c690>

In [8]: gamma.asnumpy()
Out[8]: 
array([ 1.,  1.,  1.,  1.,  1.,  1., ...............

Is this the cause for the validation low? Why the gamma changed after one epoch training?!

**Comment by winstywang** [comment]
@George19890617 This is caused by the fix_gamma parameter in BN layer. Check it carefully

**Comment by winstywang** [comment]
@hoaphumanoid Please check the latest pre-trained model:　https://github.com/dmlc/mxnet/pull/3506

**Comment by George19890617** [comment]
@winstywang thanks so much! 
So the problem is fixed just set fix_gamma back to False? I will try it.
I read the batch norm papers  https://arxiv.org/pdf/1502.03167v3.pdf
I think gamma is to learned, so when I see fix_gamma parameter, really confused at it in batch_norm-inl.h. Can you give some ideas so I can better understand it.

**Comment by hoaphumanoid** [comment]
Training now. I see a big difference in the speed: 
```bash
2016-10-12 10:24:24,979 Node[0] Start training with [gpu(0), gpu(1), gpu(2), gpu(3)]
2016-10-12 10:30:37,371 Node[0] Epoch[0] Batch [50]     Speed: 28.43 samples/sec        Train-accuracy=0.002361
2016-10-12 10:30:37,372 Node[0] Epoch[0] Batch [50]     Speed: 28.43 samples/sec        Train-top_k_accuracy_5=0.009722
2016-10-12 10:30:37,372 Node[0] Epoch[0] Batch [50]     Speed: 28.43 samples/sec        Train-top_k_accuracy_10=0.019167
2016-10-12 10:30:37,372 Node[0] Epoch[0] Batch [50]     Speed: 28.43 samples/sec        Train-top_k_accuracy_20=0.036111
```
@George19890617 have you notice the same thing?

**Comment by George19890617** [comment]
@hoaphumanoid 
the speed seem slower... But I am trying to do more experiments to diagnose the causes. 
However, I used a small data to test on inception bn full and got both good acc on train and val, So I think the problem can be solved by set fix_gamma false, 

>3518 2016-10-13 08:51:29,291 Node[0] Epoch[0] Batch [43850]  Speed: 28.01 samples/sec    Train-accuracy=0.303750
3519 2016-10-13 08:51:29,291 Node[0] Epoch[0] Batch [43850]  Speed: 28.01 samples/sec    Train-top_k_accuracy_5=0.533750
3520 2016-10-13 08:51:29,291 Node[0] Epoch[0] Batch [43850]  Speed: 28.01 samples/sec    Train-top_k_accuracy_10=0.626250
3521 2016-10-13 08:51:29,291 Node[0] Epoch[0] Batch [43850]  Speed: 28.01 samples/sec    Train-top_k_accuracy_20=0.693750
3522 2016-10-13 08:51:53,740 Node[0] Update[43871]: Change learning rate to 1.74339e-01
3523 2016-10-13 08:52:00,746 Node[0] Epoch[0] Resetting Data Iterator
3524 2016-10-13 08:52:00,746 Node[0] Epoch[0] Time cost=51921.914
3525 2016-10-13 08:52:01,005 Node[0] Saved checkpoint to "vdianTest-0001.params"
3526 2016-10-13 11:34:32,119 Node[0] Epoch[0] Validation-accuracy=0.308347
3527 2016-10-13 11:34:32,119 Node[0] Epoch[0] Validation-top_k_accuracy_5=0.542124
3528 2016-10-13 11:34:32,119 Node[0] Epoch[0] Validation-top_k_accuracy_10=0.633410
3529 2016-10-13 11:34:32,119 Node[0] Epoch[0] Validation-top_k_accuracy_20=0.710846

**Comment by hoaphumanoid** [comment]
It's been a day and I'm in iteration 1. I will leave it for more time to check if the accuracy is not decreasing. Tell me if you want me to check something.

**Comment by hoaphumanoid** [comment]
The accuracy has not decreased. I think it is ok. Feel free to open the issue again if you find any other problem.
